app/build.gradle

```diff
@@ -22,16 +24,43 @@ android {
             proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
         }
     }
+    buildFeatures {
+        dataBinding true
+    }
 }
 
 dependencies {
+    api 'com.google.android.material:material:1.3.0-alpha02'
```

Comment:
how come bringing this in via `api`  instead of `implementation`

Reply:
I copied it from a google tutorial app. I wasn't sure myself the difference between api and implementation?

Reply:
good question! an android app can contain many modules. `app` is just the name of the default module where the application lives. but think of a module like a folder where you can have code live - a larger android app might have 10,20, 100, 500+ modules. a very simple example that I tend to have in most of my app are three modules, an `app` module, a `network` module where my networking code lives, and a `database` module where most of my database operations occur. 

the benefit of having all of these modules, especially in a large code base, is you do not have to recompile the entire code base every time you make a change on one screen. in my 3 module setup for example, if i just make changes to the `network` module, I do not have to recompile the `database` module.

So in this type of multi module setup, when you bring in dependencies into each module, you can usually do it in 2 ways, via `api` or via `implementation`. When you bring in a dependency via `implementation` you are saying i don't want any module that depends on me to see this dependency.  So for example let's say we have module A, B, C, and D

if your dependency graph looked something like

A depends on B, B depends on C, C depends on D 

and you brought each one of those in via implementation, module A knows nothing about module C. And that's good! module A doesn't need to know how module B does its business.

But in that same scenario, if you brought each module into the other via `api`, the above statement would not be true. Module A would know about module C. which can be a good thing if that's the way your project is setup.

But all of that to say in a single module app, we don't need to use `api` for the time being :)

Reply:
Awesome! That makes sense. And when you talk about modules, you're talking higher level than the network package or the database package I may have in the app module?

Reply:
right, the network package is just a folder in your app module. you can extract that network folder in your app module to its own network module and have your app module bring in your network module.

Here is a sample base project I created that allows one to have features live in their own module and is a pretty good reference for a small project to have good modularization https://github.com/j-roskopf/BaseAndroidProject

for these types of projects, it's usually overkill to go all out on modularization and doesn't really matter until you get pretty large projects